python/cucumber_expressions/expression_factory.py

```diff
@@ -0,0 +1,39 @@
+import re
+
+from cucumber_expressions.expression import CucumberExpression
+from cucumber_expressions.parameter_type_registry import ParameterTypeRegistry
+from cucumber_expressions.regular_expression import RegularExpression
+
+CURLY_BRACKET_PATTERN = re.compile(r"{(.*?)}")
+INVALID_CURLY_PATTERN = re.compile(r"^\d+(?:,\d+)?$")
+
+
+class ExpressionFactory:
+    def __init__(
+        self, parameter_type_registry: ParameterTypeRegistry = ParameterTypeRegistry()
+    ):
+        self.parameter_type_registry = parameter_type_registry
+
+    @staticmethod
+    def _has_curly_brackets(string: str) -> bool:
+        return "{" in string and "}" in string
+
+    @staticmethod
+    def _extract_text_in_curly_brackets(string: str) -> list:
+        return CURLY_BRACKET_PATTERN.findall(string)
+
+    def is_cucumber_expression(self, expression_string: str):
```

Comment:
I don't think this check is simple enough. The primary constraint is explaining to people what is and is not a cucumber expression. For Java I eventually settled on requiring that all regular expressions start with `^` or end with `$` and that everything else is a Cucumber expression. This is both simple and unambiguous.

Reply:
Is it worth standardising this check then across all flavours? I have no idea what we do in ruby as I've not dug into this stuff since the initial release some 4/5 years ago

Reply:
The problem is that I cannot think of an easy way to distinguish a regex from a normal string in Python - the ^$ syntax isn't used, and are generally just strings.

Hence, I thought to try and identify the other way around - seeing if it's a Cucumber Expression.

I was a bit stuck here as I couldn't think of a reliable deterministic manner to distinguish between the two types, so input very welcome!

Reply:
I think that you can still look for the `^`/ `$` indicators.
This is a nice rule that deterministically distinguishes cucumber-expressions from a regexp ones.

> the ^$ syntax isn't used, and are generally just strings.
Pytest-BDD and other python BDD frameworks, **when no specific parser is specified (default)**, they should check if `^`/ `$` are present and if so treat it as a regular expression. If not, treat it as a cucumber-expression.
This type of regexp step-def is different from using `parsers.re`, but I think it could be a replacement for it all together.